### PR TITLE
Added NNC lowering example

### PR DESCRIPTION
**Pull request by Chillee**
Taken from https://github.com/pytorch/pytorch/pull/51130

Probably still need to add comments, but this is an annoying amount of code for a meaningfully commented example. Most of the lowering rules are obvious (about what they do), but it's still structured strangely.

### CODE REVIEW

File: fx/nnc_compile.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+# we return a somewhat hacky function that passes in the module attributes to
+# the NNC callable.
+#
+# Results
+######################################
+# Using NNC (which compiles directly to LLVM), we can compile a fairly small PyTorch model and compare performnance between NNC, PyTorch Eager, and Static Runtime.
+#
+# NNC time:  0.0066373348236083984
+# PyTorch time 0.025979042053222656
+# Static Runtime time 0.011004209518432617
```
Comment (jamesr66a) [inline fx/nnc_compile.py:42]: Maybe put your `lscpu` in a gist so people know what machine these results come from

File: fx/nnc_compile.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+# Compiling FX models to NNC (Neural Network Compiler)
+######################################################
+# The goal of this file is to demonstrate an end to end example of using FX to
+# lower a PyTorch model to a backend codegen compiler. In this example, we will
+# be using NNC
+# (https://github.com/pytorch/pytorch/blob/master/test/cpp/tensorexpr/tutorial.cpp).
+# If you're unfamiliar with NNC, the general design is strongly inspired by TVM
+# and Halide.
+#
+# To do so, this example contains two FX passes.
```
Comment (jamesr66a) [inline fx/nnc_compile.py:13]: nit: s/FX passes/FX transformations/

File: fx/nnc_compile.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+# The first one is a decomposition pass that normalizes and decomposes PyTorch
+# operations (such as addmm). Using a pass like this allows us to reduce the
+# amount of lowerings we need to write. Instead of needing to specifically
```
Comment (jamesr66a) [inline fx/nnc_compile.py:16]: `number of lowerings`

File: fx/nnc_compile.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+# write a lowering for addmm, we can decompose addmm and lower its constituent
+# operations.
+# The second one is the actual lowering pass itself. In this case, we will need
+# to convert each PyTorch operation we encounter into the corresponding NNC
+# `TensorExpr`.
+#
+# Writing these two passes, `decompose` and `nnc_compile`, are fairly similar.
```
Comment (jamesr66a) [inline fx/nnc_compile.py:23]: `is fairly similar` since this is talking about a process

File: fx/nnc_compile.py (hunk 1 above)
Comment (jamesr66a) [reply fx/nnc_compile.py:42]: Also, is Static Runtime public at all? If not, remove references to it

File: fx/nnc_compile.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+
+lowering_functions[torch.bmm] = bmm_lower
+lowering_functions[torch.mm] = mm_lower
+
+
+def lower_function(node, op, nnc_args, args):
+    inp_shapes = fx.node.map_aggregate(args, lambda arg: (arg.shape, arg.dtype) if isinstance(arg, fx.Node) else None)
+    return lowering_functions[op](node.name, node.shape, inp_shapes, nnc_args)
+
+def nnc_compile(model: torch.nn.Module, example_inputs) -> torch.nn.Module:
```
Comment (jamesr66a) [inline fx/nnc_compile.py:316]: Probably want to add a docstring and comments within this function to describe the high-level process

File: fx/nnc_compile.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+for old, new in binary_decompositions:
+    decomposition_rules[old] = binary_mapping(new)
+
+def addmm_decompose(input, mat1, mat2, beta=1, alpha=1, out=None):
+    assert(out is None)
+    return beta*input + alpha*(torch.mm(mat1, mat2))
+
+decomposition_rules[torch.addmm] = addmm_decompose
+
+def decompose(model: torch.nn.Module, example_inputs) -> torch.nn.Module:
```
Comment (jamesr66a) [inline fx/nnc_compile.py:89]: Docstring + comment high level process in this function

**Review by jamesr66a — COMMENTED** [review]
Looks pretty good

File: fx/nnc_compile.py (hunk 1 above)
Comment (Chillee) [reply fx/nnc_compile.py:42]: Bram has made references to it on the PyTorch dev discussion forum, so I assume it's public.

Comment (Chillee) [reply fx/nnc_compile.py:42]: I put my CPU version

**Pull request merged**